server/src/main/java/io/druid/server/coordinator/CachingCostBalancerStrategyFactory.java

```diff
@@ -41,115 +38,100 @@
-            }
-
-            @Override
-            public ServerView.CallbackAction segmentRemoved(DruidServerMetadata server, DataSegment segment)
-            {
-              clusterCostCacheBuilder.removeSegment(server.getName(), segment);
-              return ServerView.CallbackAction.CONTINUE;
-            }
+            clusterCostCacheBuilder.addSegment(server.getName(), segment);
+            return ServerView.CallbackAction.CONTINUE;
+          }
 
-            @Override
-            public ServerView.CallbackAction segmentViewInitialized()
-            {
-              initialized = true;
-              return ServerView.CallbackAction.CONTINUE;
-            }
+          @Override
+          public ServerView.CallbackAction segmentRemoved(DruidServerMetadata server, DataSegment segment)
+          {
+            clusterCostCacheBuilder.removeSegment(server.getName(), segment);
+            return ServerView.CallbackAction.CONTINUE;
           }
-      );
 
-      serverInventoryView.registerServerRemovedCallback(
-          executor,
-          server -> {
-            clusterCostCacheBuilder.removeServer(server.getName());
+          @Override
+          public ServerView.CallbackAction segmentViewInitialized()
+          {
+            initialized.set(true);
             return ServerView.CallbackAction.CONTINUE;
           }
-      );
+        }
+    );
 
-      lifecycleLock.started();
-    }
-    finally {
-      lifecycleLock.exitStart();
-    }
+    serverInventoryView.registerServerRemovedCallback(
+        executor,
+        server -> {
+          clusterCostCacheBuilder.removeServer(server.getName());
+          return ServerView.CallbackAction.CONTINUE;
+        }
+    );
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    // do nothing
   }
 
   @LifecycleStop
   public void stop()
   {
-    if (!lifecycleLock.canStop()) {
-      throw new ISE("CachingCostBalancerStrategyFactory can not be stopped");
-    }
     executor.shutdownNow();
-    // Not calling lifecycleLock.exitStop() because CachingCostBalancerStrategyFactory is not recycleable
   }
 
   @Override
   public BalancerStrategy createBalancerStrategy(final ListeningExecutorService exec)
   {
-    if (!lifecycleLock.awaitStarted(1, TimeUnit.MINUTES)) {
-      throw new ISE("CachingCostBalancerStrategyFactory is not started");
-    }
-    if (initialized) {
+    if (initialized.compareAndSet(true, true)) {
       try {
+        // Calling clusterCostCacheBuilder.build() in the same thread (executor's sole thread) where
+        // clusterCostCacheBuilder is updated, to avoid problems with concurrent updates
         CompletableFuture<CachingCostBalancerStrategy> future = CompletableFuture.supplyAsync(
             () -> new CachingCostBalancerStrategy(clusterCostCacheBuilder.build(), exec),
             executor
         );
         try {
-          return future.get(1, TimeUnit.SECONDS);
+          return future.get();
```

Comment:
why do above computation in separate thread and not in this thread when we are blocking this thread endlessly anyway ?

Reply:
Because clusterCostCacheBuilder is not thread-safe for concurrent use. And there is no way to copy this object without races, other than in the executor's thread as well. It would be just more complicated that the current solution, without benefits.